### PR TITLE
Firefly 708 remove selected images from viewer

### PR DESCRIPTION
**Pull request by deannaji**
Jira: https://jira.ipac.caltech.edu/browse/FIREFLY-708
K8s:  https://firefly-708-remove-selected-images-from-viewer.irsakudev.ipac.caltech.edu/irsaviewer/

Testing:
- Run a search ex: M17
- Click on "list view", instead of tiles view.
- A dialog with a table will show up, select some rows and then click "Remove Selected" button.

### CODE REVIEW

File: src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx [hunk 1]
```diff
@@ -3,113 +3,121 @@
...
 import {SelectInfo} from '../../tables/SelectInfo';
+import {addActionListener} from '../../api/ApiUtil';
+import {TABLE_SELECT} from '../../tables/TablesCntlr';
+
+
+const TABLE_ID = 'active-image-view-list-table';
+
+
+function addListener() {
+}
```
Comment (loitly) [inline src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx:35]: dead code.

File: src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx [hunk 2]
```diff
@@ -133,59 +141,98 @@ function makeModel(tbl_id,plotViewAry, expandedIds, oldModel) {
...
     const [model, setModel] = useState(undefined);
 
 
     useEffect(() => {
-        const oldModel= getTblById(TABLE_ID);
+        const oldModel = getTblById(TABLE_ID);
         if (!oldModel) {
-            watchTableChanges(TABLE_ID,[TABLE_LOADED, TABLE_FILTER, TABLE_FILTER_SELROW, TABLE_SORT],()=>dialogComplete(TABLE_ID));
+            watchTableChanges(TABLE_ID, [TABLE_LOADED, TABLE_FILTER, TABLE_FILTER_SELROW, TABLE_SORT], () => dialogComplete(TABLE_ID));
         }
```
Comment (loitly) [inline src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx:233]: This is the correct way to watch and response to table actions.

File: src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx [hunk 3]
```diff
@@ -133,59 +141,98 @@ function makeModel(tbl_id,plotViewAry, expandedIds, oldModel) {
...
+            getPlotViewAry(visRoot()).forEach((pv) => {
+                if (selectedRows.has(idx)) {
+                    dispatchDeletePlotView({plotId: pv.plotId});
+                }
+                idx++;
+            });
+            selectedRows.clear();
+        }
+    }
+}
```
Comment (loitly) [inline src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx:186]: For what you want to do, there is no need to watch for table changes.  To get the selectInfo of a table, simply call:  `TableUtil.js:getTblInfoById`  i.e.

```
const selectInfoCls = SelectInfo.newInstance(getTblInfoById(TABLE_ID)?.selectInfo);
selectInfoCls.getSelected().forEach( (rowIdx) => {
    const plotId = convertRowIdxToPlotId(rowIdx);    // todo:  need to implement
    dispatchDeletePlotView({plotId});
});
```

**Review by loitly — CHANGES_REQUESTED** [review]
There are several issues I find with this PR.  Please fix.

- If you sort the table, it remove the WRONG images.  The logic for mapping table's rowIdx to plotId is wrong.
- Watching for table changes is not needed, unless you want to enable/disable the `Remove` button based on whether table rows are selected.  Since, you are not doing this, you should at least show a popup when the button is clicked, but no rows were selected.
- Try disabling IntelliJ auto-format.  Over 80% of the changes in this PR is format related.  This makes it very difficult for reviewer to see the real changes.

**Comment by deannaji** [comment]
@loitly , fixed thanks please verify

**Review by ejoliet — APPROVED** [review]
Looks fixed.
(i noticed that there are files that has changed but only because of reformatting - not sure we want to consider those into the PR but its minor)

File: src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx [hunk 4]
```diff
@@ -3,113 +3,118 @@
...
-import {getTblById, processRequest,watchTableChanges} from '../../tables/TableUtil';
+import {getTblById, getTblInfoById, getSelectedData, processRequest, watchTableChanges} from '../../tables/TableUtil';
 import {getFormattedWaveLengthUnits, getPlotViewAry, getPlotViewById, isPlotViewArysEqual} from '../PlotViewUtil';
 import {PlotAttribute} from '../PlotAttribute';
 import {TABLE_LOADED, TABLE_FILTER, TABLE_FILTER_SELROW} from '../../tables/TablesCntlr';
 import {getViewerItemIds} from '../MultiViewCntlr';
 import {HelpIcon} from '../../ui/HelpIcon';
 import {SelectInfo} from '../../tables/SelectInfo';
+import {addActionListener} from '../../api/ApiUtil';
+import {TABLE_SELECT} from '../../tables/TablesCntlr';
```
Comment (loitly) [inline src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx:28]: These are not longer used..  can be removed.

File: src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx [hunk 5]
```diff
@@ -133,59 +138,76 @@ function makeModel(tbl_id,plotViewAry, expandedIds, oldModel) {
...
+        dispatchDeletePlotView({plotId});
+    });
+}
+
+
+function getSelectedPlotIds(){
+    const {tableModel, selectInfo} = getTblInfoById(TABLE_ID);
+    return selectInfo.selectAll ? tableModel.tableData.data.map((row) => row[2]) : Array.from(selectInfo.exceptions).map((idx)=>
+        tableModel.tableData.data[idx][2]);
+}
```
Comment (loitly) [inline src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx:175]: Logic is wrong.  I pretty much wrote the code needed to get selected PlotIds in my comment on your last PR.  Did you not see it?

If you do a select all, then unselect a row before remove images.. it removes the wrong images.

**Review by loitly — CHANGES_REQUESTED** [review]
Again, the many formatting changes are so confusing to review.  There was only 10 or so lines of changes, but I had to browse through hundreds of lines to get there.
Also, it does not work.  Try select all, then unselect one, then remove images.  You should also try filtering and sorting the table before removing images.  Please fix, test, then resubmit PR.

**Review by loitly — APPROVED** [review]
You still did not use the util function I asked you to.  However, your code works and that's good enough for me.

@ejoliet I found something while testing.  You decide whether or not it's a bug. 
The image grid and the table only sync up when the table icon is clicked.  Maybe, we should reset table sort/filter if another search is added.
To duplicate:
- Image Search: m31, WISE ALLWISE Atlas    (4 images)
- bring up table, filter on `w2`
- Image Search: m31, WISE ALLWISE Atlas    (4 images added)
- bring up table, the previous filter is now applied to the image grid

File: src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx [hunk 6]
```diff
@@ -133,59 +138,80 @@ function makeModel(tbl_id,plotViewAry, expandedIds, oldModel) {
...
+        dispatchDeletePlotView({plotId});
+    });
+}
+
+
+function getSelectedPlotIds(){
+    const {tableModel, selectInfo} = getTblInfoById(TABLE_ID);
+
+    return selectInfo.selectAll ?
+        tableModel.tableData.data.map(row => !selectInfo.exceptions.has(parseInt(row[7])) ? row[2] : '')
```
[the remark this answers is not in the document]
Comment (robyww) [reply src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx:175]: should be `(row) =>`

File: src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx [hunk 7]
```diff
@@ -133,59 +138,80 @@ function makeModel(tbl_id,plotViewAry, expandedIds, oldModel) {
...
+        :
+        Array.from(selectInfo.exceptions).map(idx =>
```
Comment (robyww) [inline src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx:177]: should be `(idx) =>`

File: src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx [hunk 8]
```diff
@@ -133,59 +138,80 @@ function makeModel(tbl_id,plotViewAry, expandedIds, oldModel) {
...
+            tableModel.tableData.data[idx][2])
```
Comment (robyww) [inline src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx:178]: `;` here

File: src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx [hunk 9]
```diff
@@ -3,113 +3,118 @@
...
     );
     DialogRootContainer.defineDialog('ExpandedOptionsPopup', popup);
     dispatchShowDialog('ExpandedOptionsPopup');
 }
 
 
-
-
-const [NAME_IDX,WAVE_LENGTH_UM,PID_IDX,STATUS, PROJ_TYPE_DESC, WAVE_TYPE, DATA_HELP_URL]= [0,1,2,3,4,5,6];
+const [NAME_IDX, WAVE_LENGTH_UM, PID_IDX, STATUS, PROJ_TYPE_DESC, WAVE_TYPE, DATA_HELP_URL] = [0, 1, 2, 3, 4, 5, 6];
```
Comment (robyww) [inline src/firefly/js/visualize/ui/ExpandedOptionsPopup.jsx:46]: Add another constant for row id.
```js
const [NAME_IDX, WAVE_LENGTH_UM, PID_IDX, STATUS, PROJ_TYPE_DESC, WAVE_TYPE, DATA_HELP_URL,ROW_ID] = [0, 1, 2, 3, 4, 5, 6,7];

```
```

**Review by robyww — CHANGES_REQUESTED** [review]
OK, it is all working, but it does need some code cleanup.

Also- I highlighted some stuff eslint catches. Do you have it configure with your intelliJ?

**Review by robyww — APPROVED** [review]

**Pull request merged**